### ISSUE TITLE
IDE0059 is not correct for async methods

### ISSUE THREAD

**Issue by SergeyTeplyakov** [dotnet/roslyn-analyzers#7640]
### Analyzer

**Diagnostic ID**: [IDE0059](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0059?pivots=lang-csharp-vb): `Remove unnecessary value assignment`

### Analyzer source
**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 8.0.0

### Describe the bug

A local variable assignment in async method between awaits should not generate IDE0059.

### Steps To Reproduce

```csharp
static async Task Async()
{
    object? o = "hello!";
    object? o2 = "world";
    // IDE0059: Unnecessary assignment of a variable to 'o'
    // This is correct!
    o = null;
    await Task.Yield();

    // IDE0059: Unnecessary assignment of a variable to 'o2'
    // This assignment is not corret, since we're setting a state
    // machine state to null!
    o2 = null;
    // a very long running method that potentially never finishes!
}
```

Assignment to `o2` should not generate IDE0059, becuase the assignment sets the state inside the generate state machine. And such assignment potentially could be resetting the last root that was needed to GC o2.
### Actual behavior

There should be no IDE0059 for `o2 = null`.

**Comment by SergeyTeplyakov** [comment]
Essentially, the analyzer should recognize if a local variable is an actual local variable or not. The issue is probably going to happen with iterator blocks.
